pkg/controller/registry/resolver/constraints/cel.go

```diff
@@ -0,0 +1,113 @@
+package constraints
+
+import (
+	"fmt"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/interpreter/functions"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+
+	"github.com/blang/semver/v4"
+)
+
+type Cel struct {
+	Rule string
+}
+
+type Evaluator interface {
+	Evaluate(env map[string]interface{}) (bool, error)
+}
+
+type EvaluatorProvider interface {
+	Evaluator(rule string) (Evaluator, error)
+}
+
+func NewCelEvaluatorProvider() *celEvaluatorProvider {
+	env, err := cel.NewEnv(cel.Declarations(
+		decls.NewVar("properties", decls.NewListType(decls.NewMapType(decls.String, decls.Any)))),
+		cel.Lib(semverLib{}),
+	)
+	if err != nil {
+		panic(err)
+	}
+	return &celEvaluatorProvider{
+		env: env,
+	}
+}
+
+type celEvaluatorProvider struct {
+	env *cel.Env
+}
+
+type celEvaluator struct {
+	p cel.Program
```

Comment:
should we call it `celProgram` or `program` instead of `p`? Single letter names make the code harder to read.